### PR TITLE
Add a Property Inspector

### LINKED ISSUE #7664: Add a Property Inspector
## Description
Add a property inspector that is used to display properties about a widget.  Currently this will be used by the notebook cell tools, but can be used by any main area widget to display additional properties about the widget.  

Typically a property inspector would be on the right hand side (like in Photoshop), but we will keep it on the left side for now, since there is currently nothing on the right side.  Users can choose to move it to the right using the context menu, and implementers of the extension can put it on the right if desired.

### PR DESCRIPTION
**Pull request by blink1073**
## References
Fixes #7664

## Code changes
Adds a `property-inspector` package and default implementation of the extension in `application-extension`.   The notebook extension is updated to use the inspector.

```typescript
/**
 * A property inspector interface provided when registering
 * to a property inspector provider.  Allows an owner widget
 * to set the property inspector content for itself.
 */
export interface IPropertyInspector extends IDisposable {
  /*
   * Render the property inspector content.
   *
   * If the owner widget is not the most recently focused,
   * The content will not be shown until that widget
   * is focused.
   *
   * @param content - the widget or react element to render.
   */
  render(content: Widget | React.ReactElement): void;

  /**
   * Show the property inspector panel.
   *
   * If the owner widget is not the most recently focused,
   * this is a no-op.  It should be triggered by a user
   * action.
   */
  showPanel(): void;
}

/**
 * A provider for property inspectors.
 */
export interface IPropertyInspectorProvider {
  /**
   * Register a widget in the property inspector provider.
   *
   * @param widget The owner widget whose properties will be inspected.
   *
   * ## Notes
   * Only one property inspector can be provided for each widget.
   * Registering the same widget twice will result in an error.
   * A widget can be unregistered by disposing of its property
   * inspector.
   */
  register(widget: Widget): IPropertyInspector;
}
```

## User-facing changes
If no other extensions are installed that use the side panel, the use will always see the tool icon even when no notebooks are open, with the text "No properties to inspect.".   

I attempted to only show the tab when one or more inspectors are registered, but it conflicts with our logic that allows side bar items to be moved to the opposite side.

image

image

## Backwards-incompatible changes
None.

### CODE REVIEW

**Comment by jasongrout** [comment]
The general functionality seems to work well. The notebook inspector needs some CSS/UX work:
Screen Shot 2019-12-23 at 9 15 24 AM


And perhaps the wrench icon should be updated to something to more reflect a property inspector.

File: packages/notebook-extension/src/index.ts [hunk 1]
```diff
@@ -302,7 +299,12 @@ const tools: JupyterFrontEndPlugin<INotebookTools> = {
   provides: INotebookTools,
   id: '@jupyterlab/notebook-extension:tools',
   autoStart: true,
-  requires: [INotebookTracker, IEditorServices, IStateDB],
+  requires: [
+    INotebookTracker,
+    IEditorServices,
+    IStateDB,
+    IPropertyInspectorProvider
```
Comment (afshin) [inline packages/notebook-extension/src/index.ts]: In a future PR, this should be made `optional`.

File: packages/property-inspector/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+
+export { IPropertyInspector, IPropertyInspectorProvider };
+
+/**
+ * The implementation of the PropertyInspector.
+ */
+abstract class PropertyInspectorProvider extends Widget
+  implements IPropertyInspectorProvider {
+  /**
+   * Constrcut a new Property Inspector.
```
Comment (afshin) [inline packages/property-inspector/src/index.ts]: Typo (construct)

File: packages/property-inspector/src/token.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+   * inspector.
+   */
+  register(widget: Widget): IPropertyInspector;
+}
+
+/**
+ * The property inspector provider token.
+ */
+export const IPropertyInspectorProvider = new Token<IPropertyInspectorProvider>(
+  '@sagemaker-ui:IPropertyInspectorProvider'
```
Comment (afshin) [inline packages/property-inspector/src/token.ts]: Wrong namespace :wink:

File: packages/property-inspector/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+/**
+ * A namespace for module private data.
+ */
+namespace Private {
+  /**
+   * A type alias for the actions a property inspector can take.
+   */
+  export type PropertyInspectorAction = 'content' | 'dispose' | 'showPanel';
```
Comment (afshin) [inline packages/property-inspector/src/index.ts]: I'd suggest `'showPanel'` be `'show-panel'` like message names.

File: packages/property-inspector/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+    }
+
+    /**
+     * Show the property inspector panel.
+     */
+    showPanel(): void {
+      if (this._isDisposed) {
+        return;
+      }
+      this._onAction.emit('showPanel');
```
Comment (afshin) [inline packages/property-inspector/src/index.ts]: I'd suggest `'showPanel'` be `'show-panel'` like message names.

File: packages/property-inspector/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+          this.setContent(sender.content);
+        }
+        break;
+      case 'dispose':
+        if (owner) {
+         this._tracker.remove(owner);
+         this._inspectors.delete(owner);
+        }
+        break;
+      case 'showPanel':
```
Comment (afshin) [inline packages/property-inspector/src/index.ts]: I'd suggest `'showPanel'` be `'show-panel'` like message names.

File: packages/application-extension/src/index.tsx [hunk 7]
```diff
@@ -765,6 +770,24 @@ const paths: JupyterFrontEndPlugin<JupyterFrontEnd.IPaths> = {
   provides: JupyterFrontEnd.IPaths
 };
 
+/**
+ * Initialization data for the property_inspector extension.
```
Comment (afshin) [inline packages/application-extension/src/index.tsx]: Strange docstring

File: packages/application-extension/style/base.css [hunk 8]
```diff
@@ -0,0 +1,5 @@
+/*-----------------------------------------------------------------------------
+| Copyright (c) 2014-2017, Jupyter Development Team.
```
Comment (afshin) [inline packages/application-extension/style/base.css]: Remove the dates?

File: packages/notebook-extension/src/index.ts [hunk 9]
```diff
@@ -70,13 +70,15 @@ import { ServiceManager } from '@jupyterlab/services';
...
 
-import { ReadonlyJSONObject, JSONValue } from '@lumino/coreutils';
+import { JSONValue } from '@lumino/coreutils';
 
 import { Message, MessageLoop } from '@lumino/messaging';
 
 import { Panel, Menu } from '@lumino/widgets';
 import { CommandRegistry } from '@lumino/commands';
 
+import { IPropertyInspectorProvider } from '@jupyterlab/property-inspector';
```
Comment (afshin) [inline packages/notebook-extension/src/index.ts]: Import order (I think this file's a mess, but we live in a society.)

**Pull request merged**